File: app/src/main/java/com/dougkeen/bart/activities/RouteArguments.java
package com.dougkeen.bart.activities;

import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.Nullable;

import com.dougkeen.bart.model.Station;
import com.dougkeen.bart.model.StationPair;

/** Primitive route and trip arguments shared by activity boundaries. */
public final class RouteArguments {
    public static final String ORIGIN = "routeOrigin";
    public static final String DESTINATION = "routeDestination";
    public static final String DEPARTURE_IDENTITY = "departureIdentity";
    public static final String SCREEN_MODE = "screenMode";

    public static final String MODE_SCHEDULE = "schedule";
    public static final String MODE_FOLLOWED = "followed";

    private RouteArguments() {
    }

    public static void putRoute(Intent intent, StationPair route) {
        if (intent == null || route == null) {
            return;
        }
        intent.putExtra(ORIGIN, abbreviation(route.getOrigin()));
        intent.putExtra(DESTINATION, abbreviation(route.getDestination()));
    }

    public static void putRoute(Bundle bundle, StationPair route) {
        if (bundle == null || route == null) {
            return;
        }
        bundle.putString(ORIGIN, abbreviation(route.getOrigin()));
        bundle.putString(DESTINATION, abbreviation(route.getDestination()));
    }

    public static void putTrip(Intent intent, StationPair route, String identity,
                               String mode) {
        putRoute(intent, route);
        intent.putExtra(DEPARTURE_IDENTITY, identity);
        intent.putExtra(SCREEN_MODE, mode);
    }

    @Nullable
    public static StationPair readRoute(Intent intent) {
        return readRoute(intent == null ? null : intent.getExtras());
    }

    @Nullable
    public static StationPair readRoute(Bundle bundle) {
        if (bundle == null) {
            return null;
        }
        Station origin = Station.getByAbbreviation(bundle.getString(ORIGIN));
        Station destination = Station.getByAbbreviation(bundle.getString(DESTINATION));
        return origin == null ? null : new StationPair(origin, destination);
    }

    @Nullable
    public static String readDepartureIdentity(Intent intent) {
        return intent == null ? null : intent.getStringExtra(DEPARTURE_IDENTITY);
    }

    @Nullable
    public static String readScreenMode(Intent intent) {
        return intent == null ? null : intent.getStringExtra(SCREEN_MODE);
    }

    private static String abbreviation(Station station) {
        return station == null ? null : station.abbreviation;
    }
}

File: app/src/main/kotlin/in/izyum/bart/activities/RouteArguments.kt
package `in`.izyum.bart.activities

import android.content.Intent
import android.os.Bundle
import `in`.izyum.bart.model.Station
import `in`.izyum.bart.model.StationPair

/** Primitive route and trip arguments shared by activity boundaries. */
object RouteArguments {
    const val ORIGIN = "routeOrigin"
    const val DESTINATION = "routeDestination"
    const val SELECTION_IDENTITY = "selectionIdentity"
    const val SCREEN_MODE = "screenMode"

    const val MODE_SCHEDULE = "schedule"
    const val MODE_FOLLOWED = "followed"

    @JvmStatic
    fun putRoute(intent: Intent?, route: StationPair?) {
        if (intent == null || route == null) return
        intent.putExtra(ORIGIN, route.origin.abbreviation)
        intent.putExtra(DESTINATION, route.destination?.abbreviation)
    }

    @JvmStatic
    fun putRoute(bundle: Bundle?, route: StationPair?) {
        if (bundle == null || route == null) return
        bundle.putString(ORIGIN, route.origin.abbreviation)
        bundle.putString(DESTINATION, route.destination?.abbreviation)
    }

    @JvmStatic
    fun putTrip(intent: Intent?, route: StationPair?, selectionIdentity: String?, mode: String?) {
        if (intent == null) return
        putRoute(intent, route)
        intent.putExtra(SELECTION_IDENTITY, selectionIdentity)
        intent.putExtra(SCREEN_MODE, mode)
    }

    @JvmStatic
    fun readRoute(intent: Intent?): StationPair? = readRoute(intent?.extras)

    @JvmStatic
    fun readRoute(bundle: Bundle?): StationPair? {
        if (bundle == null) return null
        val origin = Station.getByAbbreviation(bundle.getString(ORIGIN)) ?: return null
        return StationPair(origin, Station.getByAbbreviation(bundle.getString(DESTINATION)))
    }

    @JvmStatic
    fun readSelectionIdentity(intent: Intent?): String? =
        intent?.getStringExtra(SELECTION_IDENTITY)

    @JvmStatic
    fun readScreenMode(intent: Intent?): String? = intent?.getStringExtra(SCREEN_MODE)
}
